<a href="https://colab.research.google.com/github/Amit198181/JBVNL/blob/main/110__LTIS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
import os

uploaded = files.upload()

for fn in uploaded.keys():
    print('User uploaded file "{name}" with length {length} bytes'.format(
        name=fn, length=len(uploaded[fn])))

Saving LTIS_APR_AUG.xlsx to LTIS_APR_AUG.xlsx
User uploaded file "LTIS_APR_AUG.xlsx" with length 2230734 bytes


Load data: Load the data from the uploaded Excel file into a pandas DataFrame.
Identify relevant columns: Identify the columns containing maximum demand, contract demand, and date information.
Calculate excess demand: Calculate the percentage by which the maximum demand exceeds the contract demand for each entry.
Filter occurrences: Filter the data to identify instances where the excess demand is 10% or more.
Count occurrences: Count how many times the maximum demand exceeded the contract demand by 10% or more.
Determine if condition is met: Check if the count of occurrences is at least four.
Output result: Display whether the condition (exceeding by 10% or more at least four times) is met.
Finish task: Provide a summary of the analysis and the final result.

# Task
Analyze the uploaded Excel file to determine if the maximum demand exceeded the contract demand by 10% or more at least four times during the last financial year.

## Load data

### Subtask:
Load the data from the uploaded Excel file into a pandas DataFrame.


**Reasoning**:
Import the pandas library and load the Excel file into a DataFrame.



In [2]:
import pandas as pd

df = pd.read_excel('LTIS_APR_AUG.xlsx')
display(df.head())

ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0  RANCHI  RANCHI  RANCHI WEST  RT27482  11220279599            LIVE   
1  RANCHI  RANCHI  RANCHI WEST  RT28138  11220279600            LIVE   
2  RANCHI  RANCHI  RANCHI WEST  RT27405  11220279612            LIVE   
3  RANCHI  RANCHI  RANCHI WEST  RT28556  11220279628            LIVE   
4  RANCHI  RANCHI  RANCHI WEST  RT28602  11220279636            LIVE   

                               CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
0                M/S SABOO SONS PRATIK SABOO        LTIS             40.0   
1                               SUVESH KUMAR        LTIS              8.0   
2                               TAMIM ANSARI        LTIS              6.0   
3  SARITA SRIVASTAVA, W/O - S K SRIVASTAVA,         LTIS             10.0   
4                         SHWETA FABRICATION        LTIS             10.0   

  CONSUMER_TYPE  ... LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
0         URBAN  ...                    0.0                      0.0   
1         RURAL  ...                    0.0                      0.0   
2         RURAL  ...                    0.0                      0.0   
3         RURAL  ...                    0.0                      0.0   
4         RURAL  ...                    0.0                      0.0   

  BILLED_UNITS_HTS_202507 MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
0                    95.0           12.0                     0.0   
1                    30.0            5.0                     0.0   
2                    61.0            6.0                     0.0   
3                    55.0            4.0                     0.0   
4                    25.0            5.0                     0.0   

   SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
0                      0.0                    275.0            18.0   
1                      0.0                     70.0             4.0   
2                      0.0                    120.0             5.0   
3                      0.0                    150.0             6.0   
4                      0.0                    102.0             6.0   

   LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
0                     0.0                        0  
1                     0.0                        0  
2                     0.0                        0  
3                     0.0                        0  
4                     0.0                        0  

[5 rows x 30 columns]

## Identify relevant columns

### Subtask:
Identify the columns containing maximum demand, contract demand, and date information.


**Reasoning**:
Examine the column names in the DataFrame to identify columns related to maximum demand, contract demand, and date information.



In [3]:
print(df.columns)

Index(['ZONE', 'CIRCLE', 'DIVISION', 'CONSNO', 'ACCOUNT_NO', 'CONSUMER_STATUS',
       'CONSUMER_NAME', 'TARIFF_CODE', 'CONTRACTED_LOAD', 'CONSUMER_TYPE',
       'METER_READER_NAME', 'MF_KWH', 'PREPAID_FLAG', 'MTR_CONSUMP_TYPE',
       'BILLED_UNITS_HTS_202504', 'MDI_HTS_202504', 'BILLED_UNITS_HTS_202505',
       'MDI_HTS_202505', 'BILLED_UNITS_HTS_202506', 'MDI_HTS_202506',
       'LOAD_FACTOR_HTS_202506', 'SUPPLY_HOURS_HTS_202506',
       'BILLED_UNITS_HTS_202507', 'MDI_HTS_202507', 'LOAD_FACTOR_HTS_202507',
       'SUPPLY_HOURS_HTS_202507', 'BILLED_UNITS_HTS_202508', 'MDI_HTS_202508',
       'LOAD_FACTOR_HTS_202508', 'SUPPLY_HOURS_HTS_202508'],
      dtype='object')


## Calculate excess demand

### Subtask:
Calculate the percentage by which the maximum demand exceeds the contract demand for each entry.


**Reasoning**:
Calculate the percentage excess demand for each maximum demand column and store the results in new columns.



In [4]:
mdi_columns = [col for col in df.columns if 'MDI_HTS_' in col]
for mdi_col in mdi_columns:
    percentage_excess_col_name = f'PERCENT_EXCESS_{mdi_col}'
    df[percentage_excess_col_name] = ((df[mdi_col] - df['CONTRACTED_LOAD']) / df['CONTRACTED_LOAD']) * 100
display(df.head())

ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0  RANCHI  RANCHI  RANCHI WEST  RT27482  11220279599            LIVE   
1  RANCHI  RANCHI  RANCHI WEST  RT28138  11220279600            LIVE   
2  RANCHI  RANCHI  RANCHI WEST  RT27405  11220279612            LIVE   
3  RANCHI  RANCHI  RANCHI WEST  RT28556  11220279628            LIVE   
4  RANCHI  RANCHI  RANCHI WEST  RT28602  11220279636            LIVE   

                               CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
0                M/S SABOO SONS PRATIK SABOO        LTIS             40.0   
1                               SUVESH KUMAR        LTIS              8.0   
2                               TAMIM ANSARI        LTIS              6.0   
3  SARITA SRIVASTAVA, W/O - S K SRIVASTAVA,         LTIS             10.0   
4                         SHWETA FABRICATION        LTIS             10.0   

  CONSUMER_TYPE  ... SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  \
0         URBAN  ...                     0.0                    275.0   
1         RURAL  ...                     0.0                     70.0   
2         RURAL  ...                     0.0                    120.0   
3         RURAL  ...                     0.0                    150.0   
4         RURAL  ...                     0.0                    102.0   

  MDI_HTS_202508 LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  \
0           18.0                    0.0                        0   
1            4.0                    0.0                        0   
2            5.0                    0.0                        0   
3            6.0                    0.0                        0   
4            6.0                    0.0                        0   

   PERCENT_EXCESS_MDI_HTS_202504  PERCENT_EXCESS_MDI_HTS_202505  \
0                     -70.000000                     -70.000000   
1                     -60.000000                     -37.500000   
2                       3.333333                     -33.333333   
3                     -67.000000                     -70.000000   
4                     -56.000000                     -70.000000   

   PERCENT_EXCESS_MDI_HTS_202506  PERCENT_EXCESS_MDI_HTS_202507  \
0                     -70.000000                          -70.0   
1                     -37.500000                          -37.5   
2                     -16.666667                            0.0   
3                     -70.000000                          -60.0   
4                     -60.000000                          -50.0   

   PERCENT_EXCESS_MDI_HTS_202508  
0                     -55.000000  
1                     -50.000000  
2                     -16.666667  
3                     -40.000000  
4                     -40.000000  

[5 rows x 35 columns]

## Filter occurrences

### Subtask:
Filter the data to identify instances where the excess demand is 10% or more.


**Reasoning**:
Filter the DataFrame to include only rows where the percentage excess demand is 10% or more in any of the relevant columns.



In [5]:
percentage_excess_cols = [col for col in df.columns if 'PERCENT_EXCESS_MDI_HTS_' in col]
filtered_df = df[
    (df[percentage_excess_cols] >= 10).any(axis=1)
].copy()
display(filtered_df.head())

ZONE  CIRCLE     DIVISION     CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
9   RANCHI  RANCHI  RANCHI WEST    RT27349  11220279696            LIVE   
11  RANCHI  RANCHI  RANCHI WEST    RT27350  11220279725            LIVE   
15  RANCHI  RANCHI  RANCHI EAST  NTMR15369  11220280037            LIVE   
16  RANCHI  RANCHI  RANCHI WEST     B35246  11220280045            LIVE   
21  RANCHI  RANCHI  RANCHI WEST    RT20262  11220308839            LIVE   

             CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD CONSUMER_TYPE  ...  \
9             THASVIEK INC        LTIS            15.00         RURAL  ...   
11            AMIN AKHATAR        LTIS             6.00         RURAL  ...   
15             GENET STEEL        LTIS            85.00         RURAL  ...   
16             SANJAY SAHU        LTIS            10.00         RURAL  ...   
21  SRI RAM IQBAL  SHARMA         LTIS             6.15         URBAN  ...   

   SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508 MDI_HTS_202508  \
9                      0.0                   9580.0           43.0   
11                     0.0                    125.0            7.0   
15                     0.0                    503.0            4.0   
16                     0.0                   1284.0            9.0   
21                     0.0                     93.0            7.0   

   LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  \
9                     0.0                        0   
11                    0.0                        0   
15                    0.0                        0   
16                    0.0                        0   
21                    0.0                        0   

    PERCENT_EXCESS_MDI_HTS_202504  PERCENT_EXCESS_MDI_HTS_202505  \
9                      186.666667                     120.000000   
11                       6.666667                     -16.666667   
15                     -15.294118                      31.764706   
16                      50.000000                      50.000000   
21                     -25.203252                     -34.959350   

    PERCENT_EXCESS_MDI_HTS_202506  PERCENT_EXCESS_MDI_HTS_202507  \
9                      246.666667                     120.000000   
11                       0.000000                       0.000000   
15                     -95.294118                     -96.470588   
16                     -35.800000                     -22.200000   
21                     -34.959350                     -18.699187   

    PERCENT_EXCESS_MDI_HTS_202508  
9                      186.666667  
11                      16.666667  
15                     -95.294118  
16                     -10.000000  
21                      13.821138  

[5 rows x 35 columns]

## Count occurrences

### Subtask:
Count how many times the maximum demand exceeded the contract demand by 10% or more.


**Reasoning**:
Identify the percentage excess columns, count the number of times the value is 10 or greater for each row, store the count in a new column 'Excess_Demand_Count', and display the head of the dataframe.



In [6]:
percentage_excess_cols = [col for col in filtered_df.columns if 'PERCENT_EXCESS_MDI_HTS_' in col]
filtered_df['Excess_Demand_Count'] = (filtered_df[percentage_excess_cols] >= 10).sum(axis=1)
display(filtered_df.head())

ZONE  CIRCLE     DIVISION     CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
9   RANCHI  RANCHI  RANCHI WEST    RT27349  11220279696            LIVE   
11  RANCHI  RANCHI  RANCHI WEST    RT27350  11220279725            LIVE   
15  RANCHI  RANCHI  RANCHI EAST  NTMR15369  11220280037            LIVE   
16  RANCHI  RANCHI  RANCHI WEST     B35246  11220280045            LIVE   
21  RANCHI  RANCHI  RANCHI WEST    RT20262  11220308839            LIVE   

             CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD CONSUMER_TYPE  ...  \
9             THASVIEK INC        LTIS            15.00         RURAL  ...   
11            AMIN AKHATAR        LTIS             6.00         RURAL  ...   
15             GENET STEEL        LTIS            85.00         RURAL  ...   
16             SANJAY SAHU        LTIS            10.00         RURAL  ...   
21  SRI RAM IQBAL  SHARMA         LTIS             6.15         URBAN  ...   

   BILLED_UNITS_HTS_202508  MDI_HTS_202508 LOAD_FACTOR_HTS_202508  \
9                   9580.0            43.0                    0.0   
11                   125.0             7.0                    0.0   
15                   503.0             4.0                    0.0   
16                  1284.0             9.0                    0.0   
21                    93.0             7.0                    0.0   

   SUPPLY_HOURS_HTS_202508  PERCENT_EXCESS_MDI_HTS_202504  \
9                        0                     186.666667   
11                       0                       6.666667   
15                       0                     -15.294118   
16                       0                      50.000000   
21                       0                     -25.203252   

    PERCENT_EXCESS_MDI_HTS_202505  PERCENT_EXCESS_MDI_HTS_202506  \
9                      120.000000                     246.666667   
11                     -16.666667                       0.000000   
15                      31.764706                     -95.294118   
16                      50.000000                     -35.800000   
21                     -34.959350                     -34.959350   

    PERCENT_EXCESS_MDI_HTS_202507  PERCENT_EXCESS_MDI_HTS_202508  \
9                      120.000000                     186.666667   
11                       0.000000                      16.666667   
15                     -96.470588                     -95.294118   
16                     -22.200000                     -10.000000   
21                     -18.699187                      13.821138   

    Excess_Demand_Count  
9                     5  
11                    1  
15                    1  
16                    2  
21                    1  

[5 rows x 36 columns]

## Determine if condition is met

### Subtask:
Check if the count of occurrences is at least four.


**Reasoning**:
Filter the DataFrame to keep only rows where 'Excess_Demand_Count' is 4 or more, then count the number of rows in this filtered DataFrame and print a message based on the count.



In [7]:
final_filtered_df = filtered_df[filtered_df['Excess_Demand_Count'] >= 4]
count_at_least_four = len(final_filtered_df)

if count_at_least_four > 0:
    print(f"At least one consumer exceeded the contract demand by 10% or more at least four times. Number of such consumers: {count_at_least_four}")
else:
    print("No consumers exceeded the contract demand by 10% or more at least four times.")

At least one consumer exceeded the contract demand by 10% or more at least four times. Number of such consumers: 1601


## Output result

### Subtask:
Display the final filtered data containing consumers who exceeded the contract demand by 10% or more at least four times.


**Reasoning**:
Display the final filtered DataFrame containing consumers who exceeded the contract demand by 10% or more at least four times.



In [8]:
display(final_filtered_df)

ZONE   CIRCLE     DIVISION     CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
9       RANCHI   RANCHI  RANCHI WEST    RT27349  11220279696            LIVE   
31      RANCHI   RANCHI  RANCHI WEST    CN11878  11220410339            LIVE   
32      RANCHI   RANCHI  RANCHI WEST     MC6059  11220410347            LIVE   
34      RANCHI   RANCHI  RANCHI WEST    CN11807  11220412226            LIVE   
41      RANCHI   RANCHI  RANCHI EAST     S2240N  11220541540            LIVE   
...        ...      ...          ...        ...          ...             ...   
17648  DHANBAD     CHAS      LOYABAD   02KTI357  72250001305            LIVE   
17654  DHANBAD     CHAS     TENUGHAT   PTI23457  72250039132            LIVE   
17669  DHANBAD  DHANBAD    GOBINDPUR  GRIND2018  72250074167            LIVE   
17699  DHANBAD     CHAS         CHAS   BHIP0024  72250137967            LIVE   
17704  DHANBAD     CHAS      LOYABAD   01KTI360  72250149949            LIVE   

                        CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
9                        THASVIEK INC        LTIS            15.00   
31                     MAHATAB ANSARI        LTIS             6.00   
32                    MD JUBER ANSARI        LTIS             6.00   
34                      IRFAN ANSARI         LTIS             6.15   
41     TARKESHWAR MANDAL S/O P.MANDAL        LTIS            13.17   
...                               ...         ...              ...   
17648                    SANJAY KUMAR        LTIS            25.00   
17654                   M/S R S GROUP        LTIS            45.00   
17669                 MD FIROZ ANSARI        LTIS            90.00   
17699                SHANTIRAM MAHATO        LTIS             8.00   
17704              UJJWAL KUMAR GUPTA        LTIS             7.00   

      CONSUMER_TYPE  ... BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
9             RURAL  ...                  9580.0           43.00   
31            RURAL  ...                    77.0            8.00   
32            RURAL  ...                   252.0            8.00   
34            URBAN  ...                   114.0            8.00   
41           OTHERS  ...                  1156.0           18.00   
...             ...  ...                     ...             ...   
17648             O  ...                   878.0           31.38   
17654             O  ...                 10073.0           60.00   
17669             O  ...                  7642.0           99.00   
17699             O  ...                   637.0            8.98   
17704             O  ...                   822.0            8.50   

      LOAD_FACTOR_HTS_202508 SUPPLY_HOURS_HTS_202508  \
9                        0.0                       0   
31                       0.0                       0   
32                       0.0                       0   
34                       0.0                       0   
41                       0.0                       0   
...                      ...                     ...   
17648                    0.0                       0   
17654                    0.0                       0   
17669                    0.0                       0   
17699                    0.0                       0   
17704                    0.0                       0   

       PERCENT_EXCESS_MDI_HTS_202504  PERCENT_EXCESS_MDI_HTS_202505  \
9                         186.666667                     120.000000   
31                         33.333333                      33.333333   
32                          0.000000                      50.000000   
34                         30.081301                      46.341463   
41                         29.081245                       6.302202   
...                              ...                            ...   
17648                      24.000000                      28.000000   
17654                      22.222222                      22.222222   
17669                      10.000000                      55.555556   
17699   

## Summary:

### Data Analysis Key Findings

*   There are 1601 consumers whose maximum demand exceeded their contract demand by 10% or more at least four times during the last financial year based on the provided data.

### Insights or Next Steps

*   Further investigation into the consumption patterns of these 1601 consumers could reveal reasons for the frequent excess demand.
*   Consider contacting these consumers to discuss potential contract demand adjustments or energy management strategies.


## Summary:

### Data Analysis Key Findings

* There are {{count_at_least_four}} consumers whose maximum demand exceeded their contract demand by 10% or more at least four times during the last financial year based on the provided data.
* The `final_filtered_df` table displays these consumers along with their original contract demand, the count of times they exceeded the limit, and their newly calculated enhanced contract demand.

### Insights or Next Steps

* Further investigation into the consumption patterns of these {{count_at_least_four}} consumers could reveal reasons for the frequent excess demand.
* Consider contacting these consumers to discuss potential contract demand adjustments based on the calculated enhanced contract demand or energy management strategies.
* This analysis can be extended to include other financial years or different thresholds for excess demand.

## Output result

### Subtask:
Display the final filtered data containing consumers who exceeded the contract demand by 10% or more at least four times, including the calculated enhanced contract demand.

**Reasoning**:
Display the final filtered DataFrame containing consumers who exceeded the contract demand by 10% or more at least four times, now including the 'Enhanced_Contract_Demand' column.

In [13]:
display(final_filtered_df)

ZONE   CIRCLE     DIVISION     CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
9       RANCHI   RANCHI  RANCHI WEST    RT27349  11220279696            LIVE   
31      RANCHI   RANCHI  RANCHI WEST    CN11878  11220410339            LIVE   
32      RANCHI   RANCHI  RANCHI WEST     MC6059  11220410347            LIVE   
34      RANCHI   RANCHI  RANCHI WEST    CN11807  11220412226            LIVE   
41      RANCHI   RANCHI  RANCHI EAST     S2240N  11220541540            LIVE   
...        ...      ...          ...        ...          ...             ...   
17648  DHANBAD     CHAS      LOYABAD   02KTI357  72250001305            LIVE   
17654  DHANBAD     CHAS     TENUGHAT   PTI23457  72250039132            LIVE   
17669  DHANBAD  DHANBAD    GOBINDPUR  GRIND2018  72250074167            LIVE   
17699  DHANBAD     CHAS         CHAS   BHIP0024  72250137967            LIVE   
17704  DHANBAD     CHAS      LOYABAD   01KTI360  72250149949            LIVE   

                        CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
9                        THASVIEK INC        LTIS            15.00   
31                     MAHATAB ANSARI        LTIS             6.00   
32                    MD JUBER ANSARI        LTIS             6.00   
34                      IRFAN ANSARI         LTIS             6.15   
41     TARKESHWAR MANDAL S/O P.MANDAL        LTIS            13.17   
...                               ...         ...              ...   
17648                    SANJAY KUMAR        LTIS            25.00   
17654                   M/S R S GROUP        LTIS            45.00   
17669                 MD FIROZ ANSARI        LTIS            90.00   
17699                SHANTIRAM MAHATO        LTIS             8.00   
17704              UJJWAL KUMAR GUPTA        LTIS             7.00   

      CONSUMER_TYPE  ... MDI_HTS_202508  LOAD_FACTOR_HTS_202508  \
9             RURAL  ...          43.00                     0.0   
31            RURAL  ...           8.00                     0.0   
32            RURAL  ...           8.00                     0.0   
34            URBAN  ...           8.00                     0.0   
41           OTHERS  ...          18.00                     0.0   
...             ...  ...            ...                     ...   
17648             O  ...          31.38                     0.0   
17654             O  ...          60.00                     0.0   
17669             O  ...          99.00                     0.0   
17699             O  ...           8.98                     0.0   
17704             O  ...           8.50                     0.0   

      SUPPLY_HOURS_HTS_202508 PERCENT_EXCESS_MDI_HTS_202504  \
9                           0                    186.666667   
31                          0                     33.333333   
32                          0                      0.000000   
34                          0                     30.081301   
41                          0                     29.081245   
...                       ...                           ...   
17648                       0                     24.000000   
17654                       0                     22.222222   
17669                       0                     10.000000   
17699                       0                      0.000000   
17704                       0                           NaN   

       PERCENT_EXCESS_MDI_HTS_202505  PERCENT_EXCESS_MDI_HTS_202506  \
9                         120.000000                     246.666667   
31                         33.333333                      33.333333   
32                         50.000000                      66.666667   
34                         46.341463                      30.081301   
41                          6.302202                      13.895216   
...                              ...                            ...   
17648                      28.000000                      28.000000   
17654                      22.222222                      28.888889   
17669  

In [11]:
print(final_filtered_df[mdi_columns].dtypes)

MDI_HTS_202504                   float64
MDI_HTS_202505                   float64
MDI_HTS_202506                   float64
MDI_HTS_202507                   float64
MDI_HTS_202508                   float64
PERCENT_EXCESS_MDI_HTS_202504    float64
PERCENT_EXCESS_MDI_HTS_202505    float64
PERCENT_EXCESS_MDI_HTS_202506    float64
PERCENT_EXCESS_MDI_HTS_202507    float64
PERCENT_EXCESS_MDI_HTS_202508    float64
dtype: object


In [15]:
final_filtered_df['Condition_Met'] = final_filtered_df['Excess_Demand_Count'] >= 4
display(final_filtered_df[['CONSUMER_NAME', 'CONTRACTED_LOAD', 'Excess_Demand_Count', 'Enhanced_Contract_Demand', 'Condition_Met']].head())

/tmp/ipython-input-80165700.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_filtered_df['Condition_Met'] = final_filtered_df['Excess_Demand_Count'] >= 4


CONSUMER_NAME  CONTRACTED_LOAD  Excess_Demand_Count  \
9                     THASVIEK INC            15.00                    5   
31                  MAHATAB ANSARI             6.00                    5   
32                 MD JUBER ANSARI             6.00                    4   
34                   IRFAN ANSARI              6.15                    4   
41  TARKESHWAR MANDAL S/O P.MANDAL            13.17                    4   

    Enhanced_Contract_Demand  Condition_Met  
9                 185.000000           True  
31                 33.333333           True  
32                 54.166667           True  
34                 34.146341           True  
41                 26.310934           True

In [17]:
percentage_excess_cols = [col for col in final_filtered_df.columns if 'PERCENT_EXCESS_MDI_HTS_' in col]

for col in percentage_excess_cols:
    final_filtered_df[f'{col}_Met_10_Percent'] = final_filtered_df[col] >= 10

display(final_filtered_df.head())

/tmp/ipython-input-2521959143.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_filtered_df[f'{col}_Met_10_Percent'] = final_filtered_df[col] >= 10
/tmp/ipython-input-2521959143.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_filtered_df[f'{col}_Met_10_Percent'] = final_filtered_df[col] >= 10
/tmp/ipython-input-2521959143.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats

ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
9   RANCHI  RANCHI  RANCHI WEST  RT27349  11220279696            LIVE   
31  RANCHI  RANCHI  RANCHI WEST  CN11878  11220410339            LIVE   
32  RANCHI  RANCHI  RANCHI WEST   MC6059  11220410347            LIVE   
34  RANCHI  RANCHI  RANCHI WEST  CN11807  11220412226            LIVE   
41  RANCHI  RANCHI  RANCHI EAST   S2240N  11220541540            LIVE   

                     CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD CONSUMER_TYPE  \
9                     THASVIEK INC        LTIS            15.00         RURAL   
31                  MAHATAB ANSARI        LTIS             6.00         RURAL   
32                 MD JUBER ANSARI        LTIS             6.00         RURAL   
34                   IRFAN ANSARI         LTIS             6.15         URBAN   
41  TARKESHWAR MANDAL S/O P.MANDAL        LTIS            13.17        OTHERS   

    ... PERCENT_EXCESS_MDI_HTS_202507  PERCENT_EXCESS_MDI_HTS_202508  \
9   ...                    120.000000                     186.666667   
31  ...                     33.333333                      33.333333   
32  ...                     66.666667                      33.333333   
34  ...                     -2.439024                      30.081301   
41  ...                     21.488231                      36.674260   

   Excess_Demand_Count Enhanced_Contract_Demand  Condition_Met  \
9                    5               185.000000           True   
31                   5                33.333333           True   
32                   4                54.166667           True   
34                   4                34.146341           True   
41                   4                26.310934           True   

    PERCENT_EXCESS_MDI_HTS_202504_Met_10_Percent  \
9                                           True   
31                                          True   
32                                         False   
34                                          True   
41                                          True   

    PERCENT_EXCESS_MDI_HTS_202505_Met_10_Percent  \
9                                           True   
31                                          True   
32                                          True   
34                                          True   
41                                         False   

    PERCENT_EXCESS_MDI_HTS_202506_Met_10_Percent  \
9                                           True   
31                                          True   
32                                          True   
34                                          True   
41                                          True   

    PERCENT_EXCESS_MDI_HTS_202507_Met_10_Percent  \
9                                           True   
31                                          True   
32                                          True   
34                                         False   
41                                          True   

    PERCENT_EXCESS_MDI_HTS_202508_Met_10_Percent  
9                                           True  
31                                          True  
32                                          True  
34                                          True  
41                                          True  

[5 rows x 43 columns]

In [18]:
# Replace 'output_filename.xlsx' with your desired file name
final_filtered_df.to_excel('analyzed_demand_data.xlsx', index=False)
print("DataFrame saved to analyzed_demand_data.xlsx")

DataFrame saved to analyzed_demand_data.xlsx


## Calculate Enhanced Contract Demand

### Subtask:
For each consumer who exceeded the contract demand by 10% or more at least four times, calculate the average of their four highest maximum demand recordings.

**Reasoning**:
Select the MDI columns for the filtered consumers, find the four highest MDI values for each consumer, calculate the average of these four values, store the result in a new column 'Enhanced_Contract_Demand', and display the head of the updated DataFrame.

In [12]:
mdi_columns = [col for col in df.columns if 'MDI_HTS_' in col]

# Convert MDI columns to numeric, coercing errors
for col in mdi_columns:
    final_filtered_df[col] = pd.to_numeric(final_filtered_df[col], errors='coerce')


def calculate_enhanced_demand(row):
    mdi_values = pd.to_numeric(row[mdi_columns], errors='coerce').dropna()
    if len(mdi_values) >= 4:
        return mdi_values.nlargest(4).mean()
    return None

final_filtered_df['Enhanced_Contract_Demand'] = final_filtered_df.apply(calculate_enhanced_demand, axis=1)

display(final_filtered_df[['CONSUMER_NAME', 'CONTRACTED_LOAD', 'Excess_Demand_Count', 'Enhanced_Contract_Demand']].head())

/tmp/ipython-input-1180180126.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_filtered_df[col] = pd.to_numeric(final_filtered_df[col], errors='coerce')
/tmp/ipython-input-1180180126.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_filtered_df['Enhanced_Contract_Demand'] = final_filtered_df.apply(calculate_enhanced_demand, axis=1)


CONSUMER_NAME  CONTRACTED_LOAD  Excess_Demand_Count  \
9                     THASVIEK INC            15.00                    5   
31                  MAHATAB ANSARI             6.00                    5   
32                 MD JUBER ANSARI             6.00                    4   
34                   IRFAN ANSARI              6.15                    4   
41  TARKESHWAR MANDAL S/O P.MANDAL            13.17                    4   

    Enhanced_Contract_Demand  
9                 185.000000  
31                 33.333333  
32                 54.166667  
34                 34.146341  
41                 26.310934

In [14]:
top_10_consumers = final_filtered_df.sort_values(by='Enhanced_Contract_Demand', ascending=False).head(10)
display(top_10_consumers[['CONSUMER_NAME', 'CONTRACTED_LOAD', 'Excess_Demand_Count', 'Enhanced_Contract_Demand']])

CONSUMER_NAME  CONTRACTED_LOAD  \
17320             S RAJSHREE INDURTRIES PRO              22.00   
5247            MANISHA GUPTA,W/O MOHAN GUPTA             8.00   
14246                         SHRAWANI KUMARI             7.02   
7842                     RELIABLE VAHAN UDYOG            10.00   
16980                        VISION METCHEM               6.00   
16987    MS BOKARO PIPE FABRICATOR  ERECTOR               6.00   
14023                    RAM KUMAR MANDHYAN              12.00   
13246                     KAILASHPATI MATALIC            15.00   
17123  MAA KAMAKHYA POULTRY PRODUCT PVT. LTD.            53.00   
7268                           Kishori Devi               6.00   

       Excess_Demand_Count  Enhanced_Contract_Demand  
17320                    5                485.227273  
5247                     5                428.125000  
14246                    5                386.823362  
7842                     5                385.000000  
16980                    5                373.666667  
16987                    5                358.333333  
14023                    5                329.166667  
13246                    4                323.266667  
17123                    5                307.283019  
7268                     5                300.000000